src/daemon/distribution.c

```diff
@@ -0,0 +1,402 @@
+  for (size_t i = 0; i < d->num_buckets; ++i) {
+    counters[i] = d->buckets[i].counter;
+  }
+  pthread_mutex_unlock(&d->mutex);
+
+  return counters;
+}
+
+size_t distribution_get_num_buckets(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return 0;
+  }
+
+  return d->num_buckets;
+}
+
+double distribution_get_sum_gauges(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NAN;
+  }
+
+  return d->sum_gauges;
+}
+
+bool distribution_check_equal(distribution_t *d1, distribution_t *d2) {
+  if ((d1 == NULL && d2 != NULL) || (d1 != NULL && d2 == NULL)) {
+    return false;
+  }
+
+  /* thanks to previous condition we know that if d1 is NULL then d2 is NULL too
+   */
+  if (d1 == NULL) {
+    return true;
+  }
+
+  pthread_mutex_lock(&d1->mutex);
+  pthread_mutex_lock(&d2->mutex);
+
+  if (d1->sum_gauges != d2->sum_gauges) {
+    return false;
+  }
+
+  if (d1->num_buckets != d2->num_buckets) {
+    return false;
+  }
+
+  for (size_t i = 0; i < d1->num_buckets; ++i) {
+    if (d1->buckets[i].max_boundary != d2->buckets[i].max_boundary ||
+        d1->buckets[i].counter != d2->buckets[i].counter) {
+      return false;
+    }
+  }
+
+  pthread_mutex_unlock(&d2->mutex);
+  pthread_mutex_unlock(&d1->mutex);
+
+  return true;
+}
+
+static bucket_t *bucket_new_linear(size_t num_buckets, double size) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = (double)(i + 1) * size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_exponential(size_t num_buckets, double base,
+                                        double factor) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double multiplier = 1.0;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = factor * multiplier;
+    multiplier *= base;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_custom(size_t num_boundaries,
+                                   const double *custom_buckets_boundaries) {
+  bucket_t *buckets = calloc(num_boundaries + 1, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  if (num_boundaries > 0) {
+    if (custom_buckets_boundaries[0] <= 0 ||
+        custom_buckets_boundaries[0] == INFINITY) {
+      free(buckets);
+      errno = EINVAL;
+      return NULL;
+    }
+
+    buckets[0].max_boundary = custom_buckets_boundaries[0];
+
+    for (size_t i = 1; i < num_boundaries; ++i) {
+      if (custom_buckets_boundaries[i] <= 0 ||
+          custom_buckets_boundaries[i] == INFINITY ||
+          custom_buckets_boundaries[i - 1] >= custom_buckets_boundaries[i]) {
+        free(buckets);
+        errno = EINVAL;
+        return NULL;
+      }
```

Comment:
I'd prefer to check boundaries first and then to allocate memory for the buckets because memory allocation is very expensive operation and I'd like to avoid doing it when unnecessary

Reply:
Great idea, I will change it! Thanks!